### PR TITLE
[DI] Remove useless state from ServiceLocator

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no (master only)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

One less state to manage for the engine, and allows to deal with non-shared services.

### CODE REVIEW

**Review by chalasr — APPROVED** [review]
👍

File: src/Symfony/Component/DependencyInjection/ServiceLocator.php [hunk 1]
```diff
@@ -53,13 +52,12 @@ public function get($id)
             throw new ServiceCircularReferenceException($id, array($id, $id));
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/ServiceLocator.php]: Do we really need to cover this? The ServiceLocator is not passed to the factories so it is very unlikely to happen.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ServiceLocator.php]: Why should we remove the safe guard? Symfony is used into the wild, it needs to handle properly any situation. I see no reason to trade a few lines vs a fatal error, however rare it is.

Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/ServiceLocator.php]: Ok, just asking, fine for me as is then.

**Comment by xabbuh** [comment]
👍 

Status: Reviewed

File: src/Symfony/Component/DependencyInjection/Tests/ServiceLocatorTest.php [hunk 2]
```diff
@@ -40,15 +40,20 @@ public function testGet()
         $this->assertSame('baz', $locator->get('bar'));
     }
 
-    public function testGetDoesNotExecuteTheSameCallableTwice()
+    public function testGetDoesNotMemoize()
```
Comment (jakzal) [inline src/Symfony/Component/DependencyInjection/Tests/ServiceLocatorTest.php]: Memoize => Memorize

**Comment by jakzal** [comment]
apart from the typo :+1:

File: src/Symfony/Component/DependencyInjection/Tests/ServiceLocatorTest.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/ServiceLocatorTest.php]: Not a type :)
https://en.wikipedia.org/wiki/Memoization

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**